completed first re-factored version with passing tests

TMaile opened the pull request:
this is our first version that passed our 3 tests for now

TobiasMaile committed:
adding the validation function

TobiasMaile committed:
adding the translator classes and first tests

TobiasMaile committed:
adding facility, site, building, buildingsystem classes

mrigeshgupta committed:
Implement and test "should have a version“ test in BuildingSync
Implement and test "should parse a phase zero XML“ test

mrigeshgupta committed:
Port code from bricr project so this test runs successfully

mrigeshgupta committed:
Port code from bricr project so this test runs successfully

mrigeshgupta committed:
Run test code

mrigeshgupta committed:
Run test code

TobiasMaile committed:
adding helper files, more refactoring

mrigeshgupta committed:
Run test code

TobiasMaile committed:
refactoring of the first measure partially done

TobiasMaile committed:
add some details for Mrigesh

TobiasMaile committed:
finised inital refactoring of the first measure

mrigeshgupta committed:
Run test code

mrigeshgupta committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

# Conflicts:
#	lib/buildingsync/model_articulation/building.rb
#	lib/buildingsync/model_articulation/spatial_element.rb

TobiasMaile committed:
fixing class name mixup

TobiasMaile committed:
more cleanup

TobiasMaile committed:
fixing the issue with the xml snippets

TobiasMaile committed:
using REXML::Elements and renaming the read_xml function

TobiasMaile committed:
removing the measures that we replace

TobiasMaile committed:
renaming some functions

mrigeshgupta committed:
Resolve xml parsing issue.

mrigeshgupta committed:
Merge remote-tracking branch 'origin/DA' into DA

TobiasMaile committed:
cleanup of floor area parameters

TobiasMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

mrigeshgupta committed:
Resolve xml parsing issue.

TobiasMaile committed:
creating a num_stories function in building

TobiasMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

mrigeshgupta committed:
Resolve xml parsing issue.

mrigeshgupta committed:
Merge remote-tracking branch 'origin/DA' into DA

TobiasMaile committed:
fixing the remaining issues, tests are running now

TobiasMaile committed:
changing method names to use read prefix

TobiasMaile committed:
bug fixes

TobiasMaile committed:
bug fixes so it parses the tests and writes an OSM

TobiasMaile committed:
add reasonability checks after measure

TobiasMaile committed:
cleanup

TobiasMaile committed:
more cleanup

TobiasMaile committed:
more cleanup

TMaile committed:
adding additional test files

mrigeshgupta committed:
Resolve xml parsing issue.

mrigeshgupta committed:
Merge remote-tracking branch 'origin/DA' into DA

TMaile committed:
update on checking for 1 facility, 1 site, 1 building

TMaile committed:
adding the calibration function

TMaile committed:
enhancements for GSA test file

TMaile committed:
adding more test files

TMaile committed:
fix issue with floor area

TMaile committed:
adding logging and cleanup

TMaile committed:
adding some logging

mrigeshgupta committed:
Implement Logging

tijcolem committed:
adding Jenkinsfile for CI

macumber committed:
Adding license text

macumber committed:
Add BuildingSync trademark to copyright notices

TMaile committed:
adding more helper files, weather files and update on weather stuff

TMaile committed:
udpate in the system refactoring

nllong committed:
Merge pull request #6 from BuildingSync/pr-jenkins-ci

adding Jenkinsfile for CI

nllong commented on Rakefile:
```diff
@@ -1,52 +1,123 @@
-#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
-#  IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-########################################################################################################################
+require 'bundler'
```
We updated the Copyrights in the develop branch. Make sure to merge develop up sooner than later.

TMaile replied:
Develop branch merged into DA branch. [related commit](https://github.com/BuildingSync/BuildingSync-gem/commit/f68a2dd3f2e448eec74ad55f4c7f440adc1f4855)
Updated/added copyright in all files. [related commit](https://github.com/BuildingSync/BuildingSync-gem/commit/d5e8144a3f6e04493f370dbd7b25eb853e0f2a2c)

nllong commented on config.rb.in:
```diff
@@ -0,0 +1,63 @@
+  end
+
+  # one or more measure paths
+  OPENSTUDIO_MEASURES = [
```
we'll want to make these more configurable. Right now if these change the user will need to fork the gem or overload Module variables...

macumber replied:
This stuff is all covered by the extension gem.  Paths to measures in this gem are defined in the Extension class.  Paths from other extension gems are automatically discovered.

TMaile replied:
removed the local constant and using the extension gem now. 
@nllong anything else we can to make these more con configurable? A user configuration file, or some additional parameters to the main entry functions?

nllong commented on lib/buildingsync.rb:
```diff
@@ -1,80 +0,0 @@
-########################################################################################################################
```
is there a reason this file has moved or disappeared?

nllong commented on lib/buildingsync/Helpers/calibrate_factors.json:
```diff
@@ -0,0 +1,140 @@
+{
```
the directory Helpers needs to be lowercased.

nllong commented on lib/buildingsync/Helpers/epw.rb:
```diff
@@ -0,0 +1,229 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *******************************************************************************
+
+######################################################################
```
as mentioned above... the copyrights will be need to be cleaned up

nllong commented on lib/buildingsync/Helpers/os_lib_geometry.rb:
```diff
@@ -0,0 +1,1022 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *******************************************************************************
+
+module OsLib_Geometry
```
I am thinking that we will want to reference this module by inhereting from this gem (openstudio-model-articulation-gem)... https://github.com/NREL/openstudio-model-articulation-gem/blob/008af2a7ad7d1cce7c09c01835f57f5f7bf77180/lib/measures/SimplifyGeometryToSlicedBar/resources/os_lib_geometry.rb

nllong commented on lib/buildingsync/Helpers/os_lib_helper_methods.rb:
```diff
@@ -0,0 +1,399 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *******************************************************************************
+
+module OsLib_HelperMethods
```
also in openstudio-model-articulation-gem

nllong commented on lib/buildingsync/makers/workflow_maker_phase_zero.rb:
```diff
@@ -0,0 +1,906 @@
+      configureForDoc(@workflow)
+    end
+
+    def configureForDoc(osw)
```
should be snake_case

nllong commented on lib/buildingsync/makers/workflow_maker_phase_zero.rb:
```diff
@@ -0,0 +1,906 @@
+          built_year = major_remodel_year if major_remodel_year > built_year
+        end
+
+        if built_year < 1978
```
this appears to be very California centric. We need to make sure we abstract out this logic for ASHRAE climate zones.

macumber replied:
Agree

TMaile replied:
see #9

nllong commented on lib/buildingsync/makers/workflow_maker_phase_zero.rb:
```diff
@@ -0,0 +1,906 @@
+      end
+    end
+
+    def configureForScenario(osw, scenario)
```
snake_case

nllong commented on lib/buildingsync/model_articulation/building.rb:
```diff
@@ -0,0 +1,836 @@
+require_relative '../Helpers/epw'
+require_relative '../Helpers/stat_file'
+module BuildingSync
+  class Building < SpecialElement
```
Did you mean SpatialElement?

nllong commented on lib/buildingsync/model_articulation/spatial_element.rb:
```diff
@@ -0,0 +1,121 @@
+
+module BuildingSync
+  # base class for objects that will configure workflows based on building sync files
+  class SpecialElement
```
I think this is supposed to be SpatialElement

nllong commented on lib/buildingsync/translator.rb:
```diff
@@ -0,0 +1,72 @@
+module BuildingSync
+  class Translator
+    # load the building sync file and chooses the correct workflow
+    def initialize(xmlFilePath, outputDir, bWorkflow)
```
arguments need to be snake_cased

nllong commented on lib/buildingsync/workflow_maker.rb:
```diff
@@ -0,0 +1,70 @@
+      @ns = ns
+    end
+
+    def writeOSWs(dir)
```
snake_case... I'm only tagging these issues every once in awhile. Typically the logic is a follows. Ruby methods should be snake_case as well as variables. The only time a lowerCamelCase method should be in Ruby is if the method is in the OpenStudio C++ code itself.

TMaile replied:
changed all methods and variables to be snake_case

nllong reviewed (changes requested):
This is looking good! In general:
* lot of little items around style. 
* a lot of the helper methods should be inherited from the OpenStudio-modelarticulation-gem
* I'm not fond of having the weather files in this Gem. I don't have a better idea at the moment. In python there is a post-install command that runs and I'm wondering if we could do something similar where upon installing the gem a post install command runs and installed the weather files.

TMaile committed:
update log messages in facility and building

TMaile committed:
Merge branch 'develop' into DA

TMaile committed:
added/updated copyright statement to every file

TMaile committed:
Helpers -> helpers

TMaile committed:
remove calibration factors and update copyrights of epw.rb file

TMaile committed:
fix typo with spatial element

macumber commented on Rakefile:
```diff
@@ -1,52 +1,123 @@
 
 require 'rubocop/rake_task'
-RuboCop::RakeTask.new
+desc 'Run RuboCop on the lib directory'
```
@TMaile we should be pulling in rake tasks from the extension gem (as we are here https://github.com/NREL/openstudio-model-articulation-gem/blob/develop/Rakefile#L44) rather than redefining them.  Note that we have restructured all of our extension gems.  The main file, `lib/buildingsync.rb`, should just include other files in the `lib/buildingsync/` directory.  The extension class should be renamed `BuildingSync::Extension` and moved to `lib/buildingsync/extension.rb`.

@nllong FYI, hope we're not giving conflicting input here

TMaile replied:
this is completed as far as I can tell now

macumber commented on Rakefile:
```diff
@@ -1,52 +1,123 @@
+  task.fail_on_error = false
+end
+
+desc 'Update copyright on files'
```
This is already a rake task in the extension gem https://github.com/NREL/openstudio-extension-gem/blob/develop/lib/openstudio/extension/rake_task.rb#L129

macumber commented on Rakefile:
```diff
@@ -1,52 +1,123 @@
+end
+
+
+desc 'Init config'
```
I don't think we need this as I don't think we should be spinning up a SEED instance in the tests, @nllong is that right?  We should disable the Travis tests in preference for the Jenkins ones that @tijcolem has set up.  Basically, there are predefined Jenkins tests for all extension gems, we can update them in one place.

macumber commented on config.rb.in:
```diff
@@ -0,0 +1,63 @@
+  ]
+
+  # max number of datapoints to run
+  MAX_DATAPOINTS = Float::INFINITY
```
Some of these number of simulation variables are not yet covered by the extension gem but they should be.

TMaile replied:
I created a new issue for this #13

macumber commented on lib/buildingsync/helpers/Model.hvac.rb:
```diff
@@ -0,0 +1,210 @@
+class OpenStudio::Model::Model
```
@TMaile this seems like it should be in OpenStudio Standards, did you write this from scratch or get it somewhere else?  We should not be adding code to the `OpenStudio::Model` module in this gem, all code specific to this gem should go into the `BuildingSync` module

TMaile replied:
@macumber we got this helper file from BRICR measure `"create_typical_building_from_model_bricr`". I did not find this method `"add_cbecs_hvac_system" `with this name in any of the other gems. Did I miss it somewhere or should we just place it into `BuildingSync`.

macumber replied:
I think we should make another da specific branch of the [openstudio-model-articulation-gem](https://github.com/NREL/openstudio-model-articulation-gem/tree/develop/lib/measures/create_typical_building_from_model) to put this in.  The BuildingSync gem should not be tied to any one standard (BRICR is specific to one California based standard but BuildingSync is not).

TMaile replied:
@mrigeshgupta will do this as described in #12

mrigeshgupta committed:
Implement snake_casse codding standard

macumber commented on lib/buildingsync/helpers/epw.rb:
```diff
@@ -0,0 +1,210 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *******************************************************************************
+
+require 'csv'
```
This file is already included in the [openstudio-common-measures-gem](https://github.com/NREL/openstudio-common-measures-gem/tree/develop/lib/measures/ChangeBuildingLocation/resources).  We can make this a resource of that gem if needed.  Are you actually using this code anywhere?  It might also be possible that the [EpwFile](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/utilities/filetypes/EpwFile.hpp) class in the main OpenStudio code could do what you need?

TMaile replied:
We are using the epw.rb file for now in the openstudio-common-meaures-gem (which also adds this to the resources). 
There is a comment in the code that indicates that the OpenStudio EpwFile object can't handle multiyear weather files (or DATA PERIODS with YEARS). 
We also use the Stats.rb file in the same context.

macumber commented on lib/buildingsync/helpers/os_lib_geometry.rb:
```diff
@@ -0,0 +1,1022 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *******************************************************************************
+
+module OsLib_Geometry
```
@TMaile this seems to be a duplicate of code in the [extension gem](https://github.com/NREL/openstudio-extension-gem/blob/develop/lib/openstudio/extension/core/os_lib_geometry.rb).  You should not need to include this code in the buildingsync gem, you can just require it using `require 'openstudio/extension/core/os_lib_geometry'`

macumber commented on lib/buildingsync/helpers/os_lib_helper_methods.rb:
```diff
@@ -0,0 +1,399 @@
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# *******************************************************************************
+
+module OsLib_HelperMethods
```
Ditto all these other files, they are in the extension gem https://github.com/NREL/openstudio-extension-gem/blob/develop/lib/openstudio/extension/core/

macumber commented on lib/buildingsync/makers/workflow_maker_phase_zero.rb:
```diff
@@ -0,0 +1,917 @@
+require_relative '../workflow_maker'
+module BuildingSync
+  # base class for objects that will configure workflows based on building sync files
+  class PhaseZeroWorkflowMaker < WorkflowMaker
```
What is the difference between a ModelMaker and a WorkflowMaker, one makes a Model (OSM) and the other configures a Workflow (OSW)?  That makes sense.  We will want to create a new Measure which reads in a BuildingSync file and calls the ModelMaker to create a new OSM.

macumber replied:
When do you use ModelMaker, when do you use WorkflowMaker, and when do you use both?

TMaile replied:
@macumber the idea behind this is to distinguish between the new work (creating the OSM directly in ModelMaker) and creating the scenarios with workflows. So the ModelMaker creates the baseline and the WorkflowMaker creates the scenarios that use that baseline osm. 
To create the baseline model with ModelMaker is all you need, but if you create the scenarios then the WorkflowMaker and the ModelMaker are necessary.

macumber commented on lib/buildingsync/makers/workflow_maker_phase_zero.rb:
```diff
@@ -0,0 +1,917 @@
+        floor_area = floor_area_element.elements["#{@ns}:FloorAreaValue"].text.to_f
+        next if floor_area.nil?
+
+        floor_area_type = floor_area_element.elements["#{@ns}:FloorAreaType"].text
```
This seems like it belongs in ModelMaker.  WorkflowMaker can configure an OSW which includes a Measure that calls ModelMaker.  WorkflowMaker should not duplicate functionality of ModelMaker.

macumber commented on lib/buildingsync/model_articulation/facility.rb:
```diff
@@ -0,0 +1,493 @@
+      OpenStudio.logFree(OpenStudio::Info, 'BuildingSync.Facility.create_building_system', "The building started with #{initial_objects} objects.")
+
+      # Make the standard applier
+      standard = Standard.build(template)
```
What is Standard?  Where is it defined?  Is this an OpenStudio Standards Standard?

macumber replied:
This seems to be duplicating OpenStudio Standards code, if possible it is best to call that code directly to reuse it.

TMaile replied:
This is OpenStudio Standards Standard, so I think we are fine.

macumber commented on lib/measures/Rotate Building/measure.xml:
```diff
@@ -1,96 +1,98 @@
-   <checksum>B256482B</checksum>
-  </file>
- </files>
+<measure>
```
Rotate Building measure should probably be added to common measures and not defined here.

TMaile replied:
removed it, it is not needed right now anyway

mrigeshgupta committed:
Implement snake_casse codding standard

TMaile committed:
removing local helper files

TMaile committed:
adding OsLib_ModelGenerationBRICR again, has some changes compared to the generic one

macumber commented on lib/buildingsync/model_articulation/building.rb:
```diff
@@ -40,7 +40,7 @@
 module BuildingSync
   class Building < SpatialElement
-    include OsLib_ModelGeneration
+    include OsLib_ModelGenerationBRICR
```
@TMaile could we make a da branch of openstudio-extension-gem, push the modified version of this file there, and have you work off that branch of the extension-gem for a while?  We can follow that pattern for other repos we want to move stuff from this gem to.

TMaile committed:
argument name update

TMaile committed:
cleanup WorkflowMaker

TMaile committed:
remove the rotate building measure, it was not used anyway

TMaile committed:
cleanup of the config.rb.in

TMaile committed:
restucture of the Extension Gem usage

TMaile committed:
more rakefile changes

mrigeshgupta committed:
move "os_lib_model_generation_bricr" file to openstudio-model-articulation-gem project from BuildSync project

TMaile committed:
update gem file to include common measures and remove local epw and stat code files

TMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

TMaile committed:
fixing issues with parameters not properly passing between ModelMaker and WorkflowMaker

TMaile commented:
@nllong @macumber we addressed all comments and just have issue #9, #13 and #14 remaining. 
If there are no other objections I would merge this pull request or should I "formally" re-request a review?

TMaile committed:
additional refactoring of systems measure

mrigeshgupta committed:
implement XML validation against building schema

TMaile committed:
adding loads system

TMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

TMaile committed:
change test name and fixed typos

TMaile committed:
adding more detail to the selection tool

TMaile committed:
adding service hot water system class

TMaile committed:
thermostats and hvac refactoring

TMaile committed:
bug fixes

TMaile committed:
bug fixes

mrigeshgupta committed:
implement XML validation against building schema

mrigeshgupta committed:
Merge remote-tracking branch 'origin/DA' into DA

TMaile committed:
extracts CA climate zone in the right format now, fixes #15

TMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

macumber commented on lib/buildingsync/model_articulation/building.rb:
```diff
@@ -266,7 +266,7 @@ def set_weater_and_climate_zone(weather_file_name, climate_zone)
       elsif climate_zone == "2A" || climate_zone == "2B"
         weather_file = "CZ02RV2.epw"
-      elsif climate_zone == "3A" || climate_zone == "3B" || climate_zone == "3C"
+      elsif climate_zone == "3A" || climate_zone == "3B" || climate_zone == "3C" || climate_zone == "Climate Zone 3"
```
I think this should be coming from OS Standards somehow

TMaile replied:
This needs to be discussed. We added support to provide an EPW file directly. I have not seen a function like this in Standards.

TMaile replied:
see issue #14

mrigeshgupta committed:
implement XML validation against building schema

TMaile committed:
update testing

TMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

TobiasMaile committed:
update some names of test routines

TobiasMaile committed:
cleanup occupcany and total floor area property levels

mrigeshgupta committed:
implement XML validation against building schema

TobiasMaile committed:
improvement of the floor area handling and adding the standard to be used variable

TobiasMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

TobiasMaile committed:
enable ASHRAE 90.1 standard as an option

mrigeshgupta committed:
implement XML validation against building schema

TobiasMaile committed:
Merge branch 'DA' of https://github.com/BuildingSync/BuildingSync-gem into DA

TobiasMaile committed:
bug fixes after sizing runs works

TobiasMaile committed:
enabling the simulation run test

mrigeshgupta committed:
implement XML validation against building schema

mrigeshgupta committed:
check openstudio in unix or not

mrigeshgupta committed:
check openstudio in unix or not

mrigeshgupta committed:
check openstudio in unix or not

mrigeshgupta committed:
check openstudio in unix or not

nllong commented:
@TMaile do you think we can get this into a mergable state by our meeting next tue?

mrigeshgupta committed:
check openstudio in unix or not

TobiasMaile committed:
updates for the scenario measures

TobiasMaile committed:
adding direct EPW file and fixing issue with no subsections

TobiasMaile committed:
update the tests and when to create the standard
